File: go-backend-interview-prep/03_functions_and_methods/04_method_sets/main.go
package main

import "fmt"

// --- Interface ---
type Speaker interface {
	Speak() string
}

// --- Dog: value receiver ---
type Dog struct{ Name string }

func (d Dog) Speak() string {
	return d.Name + " says Woof"
}

// --- Cat: pointer receiver ---
type Cat struct{ Name string }

func (c *Cat) Speak() string {
	return c.Name + " says Meow"
}

// --- Mutable type with both receiver types ---
type Counter struct{ N int }

func (c Counter) Value() int { return c.N }  // value receiver
func (c *Counter) Increment()  { c.N++ }     // pointer receiver

type Getter interface{ Value() int }
type Incrementer interface{ Increment() }

func main() {
	// --- Value receiver: both T and *T satisfy the interface ---
	fmt.Println("--- Dog (value receiver) ---")
	var s Speaker

	s = Dog{Name: "Rex"}    // OK: Dog value has Speak()
	fmt.Println(s.Speak())

	s = &Dog{Name: "Buddy"} // OK: *Dog also has Speak()
	fmt.Println(s.Speak())

	// --- Pointer receiver: only *T satisfies the interface ---
	fmt.Println("\n--- Cat (pointer receiver) ---")
	s = &Cat{Name: "Whiskers"} // OK: *Cat has Speak()
	fmt.Println(s.Speak())

	// s = Cat{Name: "Fluffy"}  // COMPILE ERROR:
	// Cat does not implement Speaker (Speak has pointer receiver)

	// --- Demonstrating concrete variable auto-addressing ---
	fmt.Println("\n--- Auto-addressing on concrete variable ---")
	cat := Cat{Name: "Luna"}
	fmt.Println(cat.Speak()) // works: compiler does (&cat).Speak()
	// But you cannot do: Speaker(Cat{Name: "Luna"})

	// --- Method sets and interfaces ---
	fmt.Println("\n--- Counter method sets ---")
	c := Counter{N: 0}

	var g Getter = c     // OK: Counter has Value() (value receiver)
	fmt.Println("value:", g.Value())

	var inc Incrementer = &c // OK: *Counter has Increment() (pointer receiver)
	inc.Increment()
	inc.Increment()
	fmt.Println("after 2 increments:", c.N)

	// var inc2 Incrementer = c // COMPILE ERROR: Counter lacks Increment()
}

File: go-backend-interview-prep/03_functions_and_methods/04_method_sets/main.py
# Python equivalent of Method Sets (compare with main.go)
# Python has no value/pointer receiver distinction.
# All methods receive a reference (like Go pointer receivers).
# "Duck typing" means any object with the right method satisfies the interface.

from abc import ABC, abstractmethod


class Speaker(ABC):
    @abstractmethod
    def speak(self) -> str:
        ...


class Dog(Speaker):
    def __init__(self, name: str):
        self.name = name

    def speak(self) -> str:
        return f"{self.name} says Woof"


class Cat(Speaker):
    def __init__(self, name: str):
        self.name = name

    def speak(self) -> str:
        return f"{self.name} says Meow"


def make_speak(s: Speaker) -> None:
    print(s.speak())


def main():
    # --- No value/pointer distinction in Python ---
    # Everything is a reference. Both Dog and Cat satisfy Speaker.

    print("--- Dog ---")
    d = Dog("Rex")
    make_speak(d)

    print("\n--- Cat ---")
    c = Cat("Whiskers")
    make_speak(c)

    # In Go, Cat with pointer receiver can only satisfy Speaker as *Cat.
    # In Python, there is no such restriction -- all objects are references.

    # --- Duck typing (Python also supports this without ABC) ---
    print("\n--- Duck typing ---")

    class Duck:
        def speak(self) -> str:
            return "Quack"

    # Duck does not extend Speaker, but has speak() -- duck typing works
    duck = Duck()
    print(duck.speak())

    print("\nKey: Go distinguishes T vs *T method sets. Python does not.")


if __name__ == "__main__":
    main()
